jujugui/static/gui/src/test/test-bakery.js

```diff
@@ -0,0 +1,256 @@
+/**
+This file is part of the Juju GUI, which lets users view and manage Juju
+environments within a graphical interface (https://launchpad.net/juju-gui).
+Copyright (C) 2017 Canonical Ltd.
+
+This program is free software: you can redistribute it and/or modify it under
+the terms of the GNU Affero General Public License version 3, as published by
+the Free Software Foundation.
+
+This program is distributed in the hope that it will be useful, but WITHOUT
+ANY WARRANTY; without even the implied warranties of MERCHANTABILITY,
+SATISFACTORY QUALITY, or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Affero
+General Public License for more details.
+
+You should have received a copy of the GNU Affero General Public License along
+with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+'use strict';
+
+describe('Bakery', () => {
+  let bakery, fakeLocalStorage, macaroonlib, Y, storage, client;
+
+  beforeAll((done) => {
+    Y = YUI().use(['juju-env-bakery', 'macaroon'], (y) => {
+      Y = y;
+      macaroonlib = Y.macaroon;
+      done();
+    });
+  });
+
+  beforeEach(() => {
+    fakeLocalStorage = {
+      store: {},
+      getItem: function(item) {
+        return this.store[item];
+      },
+      setItem: function(item, value) {
+        this.store[item] = value;
+      },
+      removeItem: function(item) {
+        delete this.store[item];
+      }
+    };
+
+    client = {
+      _sendRequest: sinon.stub(),
+      sendGetRequest: function(...args) {
+        this._sendRequest('get', ...args);
+      },
+      sendPostRequest: function(...args) {
+        this._sendRequest('post', ...args);
+      },
+      sendPutRequest: function(...args) {
+        this._sendRequest('put', ...args);
+      },
+      sendDeleteRequest: function(...args) {
+        this._sendRequest('delete', ...args);
+      },
+      sendPatchRequest: function(...args) {
+        this._sendRequest('patch', ...args);
+      },
+    };
+
+    storage = new Y.juju.environments.web.BakeryStorage(
+      fakeLocalStorage, {services: {charmstore: 'http://example.com/'}});
+    bakery = new Y.juju.environments.web.Bakery(client, storage);
+  });
+
+  afterEach(() => {
+    bakery = null;
+  });
+
+  it('exists', () => {
+    bakery = new Y.juju.environments.web.Bakery({});
+    assert.isObject(bakery);
+  });
+
+  describe('can send requests', () => {
+    it('sets the method', () => {
+      bakery.sendRequest('http://example.com/', 'GET');
```

Comment:
I'd use an array for testing this, and also add more arguments? Something like
```javascript
const url = 'http://example.com/';
const headers = {header: 42};
const callback = response => 42;
let body = 'content';
let args;
['PATCH', 'post', 'PUT'].forEach(method => {
  bakery.sendRequest(url, method, header, body, callback);
  assert.equal(client._sendRequest.callCount, 1);
  args = client._sendRequest.args[0];
  assert.equal(args[0], method.toLower());
  assert.equal(args[1], url);
  assert.equal(args[2], {header: 42, 'Bakery-Protocol-Version': 1});
  assert.equal(args[3], body);
  assert.equal(args[4], bakery._wrapCallback(callback));
  client._sendRequest.reset();
}):
['GET', 'delete'].forEach(method => {
  bakery.sendRequest('http://example.com/', method, callback);
  assert.equal(client._sendRequest.callCount, 1);
  args = client._sendRequest.args[0];
  assert.equal(args[0], method.toLower());
  assert.equal(args[1], url);
  assert.equal(args[2], {header: 42, 'Bakery-Protocol-Version': 1});
  assert.equal(args[3], bakery._wrapCallback(callback));
  client._sendRequest.reset();
}):
```
What do you think?